### ISSUE TITLE
Type is incompatible with bound type for type variable

### ISSUE THREAD

**Issue by mtomassoli** [microsoft/pyright#3152]
Thanks for Pyright!

I'm using version 1.1.227.

Consider the following code:
```python
from typing import Generic, Protocol, TypeVar

K = TypeVar('K', bound='WithLessThan')

class WithLessThan(Protocol):
    def __lt__(self: K, other: K) -> bool: ...

class A(Generic[K]):
    a: K

a = A[int]
```

Here's the error report of both Pyright and mypy, for comparison:
```
>pyright bug.py
...
...\bug.py
  ...\bug.py:11:7 - error: Type "int" cannot be assigned to type variable "K@A"
    Type "int" is incompatible with bound type "WithLessThan" for type variable "K@A"
      "int" is incompatible with protocol "WithLessThan"
        "__lt__" is an incompatible type
          Type "(__x: int) -> bool" cannot be assigned to type "(other: int) -> bool"
            Position-only parameter mismatch; expected 1 but received 0 (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
Completed in 0.658sec

>mypy bug.py
Success: no issues found in 1 source file
```

The error goes away if I rename `other` into `__x`, but isn't Pyright being too rigid about this?
Who's right, Pyright or mypy?

**Comment by erictraut** [comment]
Pyright is correct in this case. The `WithLessThan` protocol requires that the `__lt__` method accept the parameter `other` by keyword, but the `__eq__` method in `int` would crash if you attempted to call it with the keyword `other`. Mypy apparently doesn't enforce this, but it should.

You can work around this by changing the `WithLessThank.__lt__` as follows:

```python
    def __lt__(self: K, other: K, /) -> bool:
        ...
```

or

```python
    def __lt__(self: K, __other: K) -> bool:
        ...
```
